model/src/application/command/option.rs

```diff
@@ -398,15 +422,47 @@ impl CommandOptionType {
             CommandOptionType::Channel => "Channel",
             CommandOptionType::Role => "Role",
             CommandOptionType::Mentionable => "Mentionable",
+            CommandOptionType::Number => "Number",
         }
     }
 }
 
+#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
+pub struct Number(pub f64);
+
+#[allow(clippy::cast_possible_truncation, clippy::trivially_copy_pass_by_ref)]
+impl Number {
+    /// Canonicalizes internal `f64` to `i64`.
+    fn canonicalize(&self) -> i64 {
+        (self.0 * 1024.0 * 1024.0).round() as i64
+    }
+}
+
+impl Eq for Number {}
+
+impl Hash for Number {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.canonicalize().hash(state);
+    }
+}
+
+impl PartialEq for Number {
+    fn eq(&self, other: &Number) -> bool {
+        self.canonicalize() == other.canonicalize()
+    }
+}
```

Comment:
IIRC the reason why `f64` doesn't implement `Eq` or `Hash` is because of `NaN`. Is `NaN` a valid value for a number choice for Discord? I'm assuming it isn't since JSON can't even represent them.

If it isn't, maybe it would be easier and more correct to have a `NotNan` type that can just have a natural implementation of `Eq` and `Hash` without worrying about `NaN` values. See [`ordered_float::NotNan`](https://docs.rs/ordered-float/2.7.0/ordered_float/struct.NotNan.html). If having an additional dependency is fine, using that type may be easiest.

This is just a suggestion though, please ignore if the current implementation is good enough.

Reply:
I think the current implementation works but I also think it can be simplified.

Canonicalizing an `f64` for a wrapper type is only important when you intent to implement `Ord` on it.
For `Hash` and `PartialEq` however, I believe a simple `f64::to_bits` is sufficient.
This way you lose the ability to compare values in size, which isn't needed anyway, and in turn you gain a tiny bit performance.

Another difference to the current implementation is that instead of *all* `NaN`s being equal, every `NaN` is only equal to itself. Both cases are distinctions from `f64` but I don't think this matters either way.